### ISSUE TITLE
Not solving the good equation

### ISSUE THREAD

**Issue by tchaton** [VahidooX/DeepLDA#1]
Hello,

I am converting this code to tensorflow even if it is a little bit complicated.
While studying the code, I found that you are solving 
evals_t = T.slinalg.eigvalsh(Sb_t, St_t)
where it should be
evals_t = T.slinalg.eigvalsh(Sb_t, Sw_t) to follow the equation.
And in the original code
from scipy.linalg.decomp import eigh
evals, evecs = eigh(Sb, Sw)
My email adress is Thomas.Chaton@uk.fujitsu.com is you want to talk about it.

**Comment by RoyiAvital** [comment]
By the way, where is the gradient calculation of the objective?
Or are you relying on Auto Grad?

**Comment by tchaton** [comment]
Yeah, we are relying on autoGrad to do it.

2018-05-25 20:16 GMT+00:00 Royi <notifications@github.com>:

**Comment by RoyiAvital** [comment]
@tchaton , Is it smooth?
As I'd assume the need for sub gradient as the objective function includes threshold.

**Comment by tchaton** [comment]
In theano it works fine. I implemented in Tensorflow, but it doesn t
support it. The loss was exploding even with matrice simplificaiton ...
I was suprised people were able to implement autoGrad over a GEP solver
ahahah

2018-05-25 20:21 GMT+00:00 Royi <notifications@github.com>:

**Comment by RoyiAvital** [comment]
Did you verify results of the Auto Grad are like in the appendix of the paper?

**Comment by tchaton** [comment]
I didn t . I assumed it was tested before hand.

2018-05-26 5:56 GMT+00:00 Royi <notifications@github.com>:

**Comment by RoyiAvital** [comment]
[It seems it hasn't been](https://github.com/CPJKU/deep_lda/issues/1#issuecomment-392250186).

**Comment by zorrocai** [comment]
@tchaton  I have written the deeplda in pytorch.But it seems that it doesn't work well too. 
Insteading of the loss exploding problem in your tensorflow version, the training problem in pytorch version is that all the eigenvalues converge to 1,and the trace ratio  matrix converges to identity matrix.
I found the reason for this phenomenon is the formula:Sw+lambda*I.
But when I remove this line code, there comes out singular matrix problem during training.
Do you have any thoughts over this problem?

**Comment by tchaton** [comment]
the Lambda*I is here to prevent det(SW) to become =0 and prevent inv. It
should work better with it. But I have no idea why it doesn't work. I had
reported the error on tensorflow, never got an answer from the team. And it
is a very important feature.

2018-05-26 13:08 GMT+00:00 zorrocai <notifications@github.com>:

**Comment by tchaton** [comment]
But yet, by looking at the paper, this code doesn t solve the correct equation ...

**Comment by zorrocai** [comment]
Thanks. there are so many hard obstacles in the deeplda model...

**Comment by tchaton** [comment]
Can you send me the pytorch code ?

2018-05-26 13:53 GMT+00:00 zorrocai <notifications@github.com>:

**Comment by zorrocai** [comment]
@tchaton Of course, sending  pytorch code to your email  Thomas.Chaton@uk.fujitsu.com.

**Comment by tchaton** [comment]
Thanks you !

2018-05-27 3:20 GMT+00:00 zorrocai <notifications@github.com>:

**Comment by dmatte** [comment]
Hi,

this is Matthias from the theano [DeepLDA repo](https://github.com/CPJKU/deep_lda).
I just wanted to comment on some of your comments.

@tchaton 
> evals_t = T.slinalg.eigvalsh(Sb_t, St_t)
> where it should be
> evals_t = T.slinalg.eigvalsh(Sb_t, Sw_t) to follow the equation.
> And in the original code
> from scipy.linalg.decomp import eigh
> evals, evecs = eigh(Sb, Sw)

Both versions should be fine. Actually, the two formulations should be equivalent.
You can find a reference on this in Section 2 of this paper:
http://mi.eng.cam.ac.uk/~cipolla/publications/inproceedings/2007-CVPR-Kim-incremental.pdf

I also just checked the two implementations on mnist and both worked in my theano implementation.

@RoyiAvital 
> Did you verify results of the Auto Grad are like in the appendix of the paper?

When we did our theano implementation and experiments we also relied on autograd.
For the paper (appendix) we wanted to also provide a sketch on how to get to the gradients of the DeepLDA objective. The actual gradients of the generalized eigenvalues that apply in DeepLDA can be found in theano#s EigvalshGrad operator: https://github.com/Theano/theano/blob/d395439aec5a6ddde8ef5c266fd976412a5c5695/theano/tensor/slinalg.py#L385-L440

This should be in line with what is provided in Section 2.1 Equation (4)
(Generalized Eigen Value Problems) of De Leeuw 2007 (http://gifi.stat.ucla.edu/speed/deleeuw_R_07c.pdf)

Regarding the gradients (and numerical stability) in PyTorch and Tensorflow
I unfortunately can't comment.
As I said, when we first tried the idea with theano everything worked fine for us from the beginning
so we did not investigate this in detail.

**Comment by zorrocai** [comment]
@tchaton @dmatte I think the problem that caused the  loss exploding problem in your tensorflow version or the poor loss in pytorch version may be the different eigenvalue decomposition methods.
there are two basic different eigen decompostion methods in both scipy and numpy:
**eig** eigenvalues and right eigenvectors for non-symmetric arrays
**eigh** Eigenvalues and right eigenvectors for symmetric/Hermitian arrays.
and they don't get the same eigenvalues, scipy's **eig** and **eigh** are totally different. numpy's **eig** and **eigh** return the same eigenvalues in different order. But numpy's **eigh** returns the different eigenvalues from scipy's **eigh**. Does this mean A\vec{e_{i}} = v_{i} B\vec{e_{i}} has the different eignevalues with \frac{A}{B}\vec{e_{i}} = v_{i}\vec{e_{i}}?
What's more I compared the torch.eig with above methods, torch.eig have the same eigenvalues as numpy.eig. Though s_b/s_w is symmetric, torch.svd retruns singular values with minus on some eigenvalues.
like this:
>>> import numpy as np
>>> import scipy
>>> import torch
>>> a = np.random.rand(1000,10)
>>> b = np.random.rand(1000,10)
>>> s_w = np.cov(a.T)
>>> s_b = np.cov(b.T)
>>> from scipy.linalg.decomp import eigh
>>> evals, evecs = eigh(s_b, s_w)
>>> evals, evecs = scipy.linalg.eigh(s_b, s_w)
>>> evals
array([ 0.82751224,  0.8575706 ,  0.90092649,  0.93457503,  0.96302211,
        1.00898617,  1.02661205,  1.10996722,  1.14103357,  1.26663491])
>>> evals, evecs = scipy.linalg.eig(s_b, s_w)
>>> evals
array([ 1.26663491+0.j,  1.14103357+0.j,  1.10996722+0.j,  0.82751224+0.j,
        0.85757060+0.j,  0.90092649+0.j,  0.93457503+0.j,  0.96302211+0.j,
        1.02661205+0.j,  1.00898617+0.j])
>>> evals, evecs = np.linalg.eigh(s_b/s_w)
>>> evals
array([-14.82691125,  -9.93629821,  -3.0339017 ,  -0.99347968,
        -0.23726914,   1.46624221,   2.76740404,   4.01231621,
        12.13940769,  18.58078932])
>>> evals, evecs = np.linalg.eig(s_b/s_w)
>>> evals
array([ 18.58078932, -14.82691125,  12.13940769,  -9.93629821,
        -3.0339017 ,   4.01231621,   2.76740404,  -0.99347968,
        -0.23726914,   1.46624221])
>>> s_w = torch.from_numpy(s_w)
>>> s_b = torch.from_numpy(s_b)
>>> evals,evecs = torch.eig(s_b/s_w)
>>> evals
tensor([[ 18.5808,   0.0000],
        [-14.8269,   0.0000],
        [ 12.1394,   0.0000],
        [ -9.9363,   0.0000],
        [ -3.0339,   0.0000],
        [  4.0123,   0.0000],
        [  2.7674,   0.0000],
        [ -0.9935,   0.0000],
        [ -0.2373,   0.0000],
        [  1.4662,   0.0000]], dtype=torch.float64)

>>> u,s,v= torch.svd(s_b/s_w)
>>> s
tensor([ 18.5808,  14.8269,  12.1394,   9.9363,   4.0123,   3.0339,
          2.7674,   1.4662,   0.9935,   0.2373], dtype=torch.float64)
>>> torch.equal((s_b/s_w).t(),s_b/s_w)
True

**Comment by tchaton** [comment]
Hello,

wow you did a great job. Did you try to check the part of my code where I
am using cholesky decomposition to simplify the equation.
I was also thinking to use Penrose_inverse in the case the matrix is not
inversible for some reasons
https://en.wikipedia.org/wiki/Moore%E2%80%93Penrose_inverse
Do you have any ideas about those ?

Best,
Tom

2018-05-29 12:41 GMT+00:00 zorrocai <notifications@github.com>:

**Comment by tchaton** [comment]
Hello,

I have found this very interesting code :
https://github.com/roatienza/Deep-Learning-Experiments/blob/master/Experiments/Tensorflow/Math/decomposition.py

Make sure to look at it :)

Best,
Tom

2018-06-05 21:34 GMT+00:00 thomas chaton <thomas.chaton.ai@gmail.com>:

**Comment by zorrocai** [comment]
Hi, @tchaton 
Yeah, for symmetric matrix,eigenvalues are the same as singular values.
![image](https://user-images.githubusercontent.com/24424749/41072600-6e77b926-6a31-11e8-941c-8c229983e128.png)
you can see it in matrix cookbook for detail.
[https://www.math.uwaterloo.ca/~hwolkowi/matrixcookbook.pdf](url)
And I found that I have made a mistake in the equation <a href="https://www.codecogs.com/eqnedit.php?latex=A\vec{e_{i}}&space;=&space;v_{i}&space;B\vec{e_{i}}" target="_blank"><img src="https://latex.codecogs.com/gif.latex?A\vec{e_{i}}&space;=&space;v_{i}&space;B\vec{e_{i}}" title="A\vec{e_{i}} = v_{i} B\vec{e_{i}}" /></a>
it should the same as <a href="https://www.codecogs.com/eqnedit.php?latex=B^{-1}A\vec{e_{i}}&space;=&space;v_{i}&space;\vec{e_{i}}" target="_blank"><img src="https://latex.codecogs.com/gif.latex?B^{-1}A\vec{e_{i}}&space;=&space;v_{i}&space;\vec{e_{i}}" title="B^{-1}A\vec{e_{i}} = v_{i} \vec{e_{i}}" /></a>, NOT <a href="https://www.codecogs.com/eqnedit.php?latex=\frac{A}{B}\vec{e_{i}}&space;=&space;v_{i}\vec{e_{i}}" target="_blank"><img src="https://latex.codecogs.com/gif.latex?\frac{A}{B}\vec{e_{i}}&space;=&space;v_{i}\vec{e_{i}}" title="\frac{A}{B}\vec{e_{i}} = v_{i}\vec{e_{i}}" /></a>. Cause that <a href="https://www.codecogs.com/eqnedit.php?latex=\frac{A}{B}=AB^{-1}" target="_blank"><img src="https://latex.codecogs.com/gif.latex?\frac{A}{B}=AB^{-1}" title="\frac{A}{B}=AB^{-1}" /></a>.
What is more, though Sw(within scatter matrix) and Sb(between scatter matrix) are both symmetric matrixes, but <a href="https://www.codecogs.com/eqnedit.php?latex=S_{b}\ast&space;S_{w}^{-1}" target="_blank"><img src="https://latex.codecogs.com/gif.latex?S_{b}\ast&space;S_{w}^{-1}" title="S_{b}\ast S_{w}^{-1}" /></a> is not symmetric.

Best,
Zorro

**Comment by RoyiAvital** [comment]
@zorrocai ,

Singular Values are the Eigen Values of $ {A}^{T} A $ or $ A {A}^{T} $ (Which share the same eigen values).
Hence the connection between the two is usually $ \sigma {\left( A \right)}_{i} = \lambda {\left( A \right)}_{i}^{2} $.

**Comment by zorrocai** [comment]
@RoyiAvital    Thanks

**Comment by nikparth** [comment]
Hi @zorrocai I am also attempting to work on the deep-lda generalized eigenvalue problem loss in pytorch so I was wondering if you might be willing to share your code as well with me. My email is nikparth@gmail.com. Thanks!

**Comment by madarax64** [comment]
Hello,
Has anyone had any luck with porting the code to PyTorch?

**Comment by magnum-zx** [comment]
@zorrocai hello, did you succeed in implementing deeplda on Pytorch, the ldaloss in my code doesn't work.......
